1.preprocessing_data/scripts/1.evaluate_qc.py

```diff
@@ -110,23 +110,86 @@
 plt.show()
 
 
+# In[5]:
+
+
+summary_statistics = df["ImageQuality_PowerLogLogSlope"].describe()
+print(summary_statistics)
+
+
+# In[6]:
+
+
+# Calculate Z-scores for the column
+z_scores = zscore(df['ImageQuality_PowerLogLogSlope'])
+
+# Set a threshold for Z-scores (adjust as needed for number of standard deviations away from the mean)
+threshold_z = 2
+
+# Identify outlier rows based on Z-scores above and below the mean since we are using absolute values of the z-scores
+blur_outliers = df[abs(z_scores) > threshold_z]
+
+# Remove any blur outliers detected in for all channels (currently we have concluded DNA and PM channels are best able to detect blurry images)
+blur_outliers = blur_outliers[(blur_outliers['Channel'] == 'DNA') | (blur_outliers['Channel'] == 'PM')]
+
+print(blur_outliers.shape)
+print(blur_outliers['Channel'].value_counts())
+blur_outliers.sort_values(by='ImageQuality_PowerLogLogSlope', ascending=False).head()
+
+
+# In[7]:
+
+
+# Calculate the mean and standard deviation
+mean_value = df["ImageQuality_PowerLogLogSlope"].mean()
+std_dev = df["ImageQuality_PowerLogLogSlope"].std()
+
+# Set the threshold multiplier for above and below the mean
+threshold = 2
+
+# Calculate the threshold values
+threshold_value_above_mean = mean_value + threshold * std_dev
+threshold_value_below_mean = mean_value - threshold * std_dev
+
+# Print the calculated threshold values
+print("Threshold for outliers above the mean:", threshold_value_above_mean)
+print("Threshold for outliers below the mean:", threshold_value_below_mean)
+
+
+# In[8]:
+
+
+sns.set_style('whitegrid')
+sns.kdeplot(data=df[df['Channel'].isin(['DNA', 'PM'])], x='ImageQuality_PowerLogLogSlope', hue='Channel', palette=['b', 'magenta'], fill=True, common_norm=False)
+plt.title(f'DNA and PM channels are best for identifying blur\n for {plate}')
+plt.xlabel('ImageQuality_PowerLogLogSlope')
+plt.ylabel('Density')
+
+plt.axvline(x=-1.2827898535807258, color='k', linestyle='--')
+plt.axvline(x=-2.3840811649380393, color='k', linestyle='--')
```

Comment:
I am assuming these raw x values are the threshold values? Is there a way to dynamically define these instead of hard coding?

Reply:
Nice catch! I do already have these as dynamic variables, I just forgot to use them! I will update this now!